Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,493 @@
+-/
+def uget (v : Vector α n) (i : USize) (h : i.toNat < n) : α :=
+  Array.uget v.toArray i (v.size_eq.symm ▸ h)
+
+
+/--
+borrowed from array API:
+auxiliary declaration used in the equation compiler when pattern matching array literals.
+-/
+abbrev getLit (v : Vector α n) (i : Nat) (h₂ : i < n) : α :=
+  Vector.get v ⟨i, h₂⟩
+
+instance : GetElem (Vector α n) Nat α fun _ i => i < n where
+  getElem := fun x i h => get x ⟨i, h⟩
+
+/--
+`pop v` returns the vector with the last element removed
+-/
+def pop (v : Vector α n) : Vector α (n-1) :=
+  ⟨Array.pop v.toArray, by {simp only [size_pop, v.size_eq]}⟩
+
+/--
+`shrink v m` removess the last `m` elements of a vector.
+-/
+def shrink (v : Vector α n) (m : Nat) : Vector α (n - m) :=
+  match m with
+  | 0 => v
+  | (x + 1) =>
+      have h : n - 1 - x = n - (x + 1) := by
+        omega
+      h ▸ (shrink (Vector.pop v) x)
+
+/--
+`isEqvAux a b hsz p i` is a helper function for `isEqv`. It performs
+an index by index check on whether the corresponding elements of a vector
+satisfy the boolean property `p` starting at a given index `i`
+-/
+def isEqvAux (a b : Vector α n)
+  (p : α → α → Bool)
+  (i : Nat) : Bool :=
+  if h : i < size a then
+     have : i < size b := h
+     p a[i] b[i] && isEqvAux a b p (i+1)
+  else
+    true
+termination_by size a - i
+decreasing_by simp_wf; decreasing_trivial_pre_omega
+
+/--
+`isEqv` takes a given boolean property `p`. It returns `true`
+if and only if `p a[i] b[i]` holds true for all valid indices `i`.
+-/
+def isEqv (a b : Vector α n) (p : α → α → Bool) : Bool := isEqvAux a b p 0
+
+instance [BEq α] : BEq (Vector α n) :=
+  ⟨fun a b => isEqv a b BEq.beq⟩
+
+
+/--`reverse v` reverses the vector `v`-/
+def reverse (v : Vector α n) : Vector α n :=
+  ⟨Array.reverse v.toArray, proof⟩
+  where
+    proof := by
+      simp
+      exact v.size_eq
+      done
+/--
+Remove the element at a given index from a vector without bounds checks, using a Fin index.
+
+This function takes worst case O(n) time because it has to backshift all elements at positions
+greater than i.
+-/
+def feraseIdx (v : Vector α n) (i : Fin n) : Vector α (n-1) :=
+  ⟨Array.feraseIdx v.toArray (v.size_eq.symm ▸ i), proof⟩
+  where
+    proof := by
+      simp [Array.size_feraseIdx]
+      rw[v.size_eq]
+
+
+/--
+  `Vector.tail` produces the tail of a vector
+-/
+def tail (v : Vector α n)  : Vector α (n-1):=
+  match n with
+  | 0 => v
+  | _ + 1 => Vector.feraseIdx v 0
+
+
+/-- Remove the element at position `i` from a vector of length `n` if `i < n`. Panics otherwise.
+
+This function takes worst case O(n) time because it has to backshift all elements at positions
+greater than i.
+ -/
+def eraseIdx! (v : Vector α n) (i : Nat) : Vector α (n-1) :=
+    if h : i < n then
+      feraseIdx v ⟨i,h⟩
+    else
+      have _ : Inhabited (Vector α (n-1)) := ⟨v.tail⟩
+      panic! "Index" ++ toString i ++ "is out of bounds"
+
+@[deprecated (since := "2024-05-04")] alias removeNth := eraseIdx
```

Comment:
Delete.

Reply:
Isn't this alias kept around for compatibility reasons?

Reply:
It's an alias in the Array namespace, `removeNth` never existed in the Vector namespace so this deprecated alias makes no sense.

Reply:
Done.